gtda/images/filtration.py

```diff
@@ -0,0 +1,150 @@
+
+    n_dimensions_ : int
+        Dimension of the images. Set in meth:`fit`.
+
+    mesh_ : int
+        Mesh image for which each pixel value is its distance to the hyperplane
+        implied by `direction_`. Set in meth:`fit`.
+
+    max_value_: float
+        Maximum pixel value among all pixels in all images of the collection.
+        Set in meth:`fit`.
+
+    See also
+    --------
+    gtda.homology.CubicalPersistence
+
+
+    """
+    _hyperparameters = {'n_dimensions_': [int, [2, 3]],
+                        'direction_': [np.ndarray, (numbers.Number, None)]}
+
+    def __init__(self, direction=None, n_jobs=None):
+        self.direction = direction
+        self.n_jobs = n_jobs
+
+    def _calculate_height(self, X):
+        Xh = np.full(X.shape, self.max_value_)
+
+        for i in range(Xh.shape[0]):
+            Xh[i][X[i] == True] = np.dot(self.mesh_[X[i] == True],
+                                         self.direction_).reshape((-1,))
+
+        return Xh
+
+    def fit(self, X, y=None):
+        """Calculate `n_dimensions_`, 'mesh_' and `max_value_` from a
+        collection of binary image. Then, return the estimator.
+
+        This method is here to implement the usual scikit-learn API and hence
+        work in pipelines.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_pixels_x, n_pixels_y [, n_pixels_z])
+            Input data. Each entry along axis 0 is interpreted as a 2D or 3D
+            grayscale image.
+
+        y : None
+            There is no need of a target in a transformer, yet the pipeline API
+            requires this parameter.
+
+        Returns
+        -------
+        self : object
+
+        """
+        X = check_array(X,  ensure_2d=False, allow_nd=True)
+
+        self.n_dimensions_ = len(X.shape) - 1
+
+        if self.direction is None:
+            self.direction_ = np.ones((self.n_dimensions_, ))
+        else:
+            self.direction_ = np.copy(self.direction)
+
+        validate_params({**self.get_params(), 'direction_': self.direction_,
+                         'n_dimensions_': self.n_dimensions_},
+                         self._hyperparameters)
+
+        self.direction_ = self.direction_ / np.linalg.norm(self.direction_)
+
+        axis_order = [2, 1, 3]
+        mesh_range_list = \
+            [np.arange(X.shape[order]) if self.direction_[i] >= 0
+             else -np.flip(np.arange(X.shape[order])) for i, order
+             in enumerate(axis_order[: self.n_dimensions_])]
+
+        self.mesh_ = np.stack(np.meshgrid(*mesh_range_list, indexing='xy'),
+                              axis=self.n_dimensions_)
+
+        self.max_value_ = 0.
+        self.max_value_ = np.max(self._calculate_height(
+            np.ones((1, *X.shape[1:]))))
+
+        return self
+
+    def transform(self, X, y=None):
+        """For each collection of binary images, calculate the corresponding
+        collection of grayscale images based on the threshold.
```

Comment:
What threshold? Also, it might be useful to have more intuition regarding what  this function is doing.

Is it correct to say that each binary image represents whether a pixel belongs to it or not? Then, each pixel that is in the image is added, starting from the origin, after a time proportional to its projection along a predefined direction. The pixels not in the image are added after all the projected_pixels.

Reply:
Threshold is from a wrong copy paste. Sorry, that was my mistake.

Your definition of a binary image is correct.
The height filtration of a binary image is a grayscale image in whose pixel values are calculated based on the distance of the pixel to an hyperplane determined by the direction. Deactivated pixels takes a pixel value corresponding to the pixel value corresponding to the maximum distance between the hyperplane and all pixels of the image + 1.

Let me rephrase that into something intelligible.